src/backend/distributed/executor/executor.c

```diff
@@ -1002,29 +1049,52 @@ FindOrCreateWorkerSession(WorkerPool *workerPool, MultiConnection *connection)
 void
 RunDistributedExecution(DistributedExecution *execution)
 {
-	int workerCount = list_length(execution->workerList);
+	int workerCount = 0;
+	int maxWaitEventCount = 0;
+	WaitEvent *events = NULL;
+	bool sequential = ShouldRunTasksSequentially(execution);
+
+	workerCount = list_length(execution->workerList);
 
 	/* additional 2 is for postmaster and latch */
-	int maxWaitEventCount = execution->totalTaskCount * workerCount + 2;
+	maxWaitEventCount = execution->totalTaskCount * workerCount + 2;
 
 	/* allocate events for the maximum number of connections to avoid realloc */
-	WaitEvent *events = palloc0(execution->totalTaskCount * workerCount *
-								sizeof(WaitEvent));
+	events = palloc0(maxWaitEventCount * sizeof(WaitEvent));
 
 	PG_TRY();
 	{
 		bool cancellationReceived = false;
 
 		execution->waitEventSet = BuildWaitEventSet(execution->sessionList);
 
-		while (execution->unfinishedTaskCount > 0 && !cancellationReceived)
+		while (execution->finishedTaskCount != execution->totalTaskCount &&
+			   !cancellationReceived)
 		{
 			long timeout = 1000;
 			int connectionCount = list_length(execution->sessionList);
 			int eventCount = 0;
 			int eventIndex = 0;
 			ListCell *workerCell = NULL;
 
+			if (execution->finishedTaskCount == execution->scheduledTaskCount)
+			{
+				/* we need to schedule more tasks */
+				List *taskListToSchedule = NIL;
+				if (sequential)
+				{
+					taskListToSchedule = list_make1(linitial(execution->unscheduledTaskList));
```

Comment:
How do we ensure that the previous task has been finished when we get back here second (or nth time) to before scheduling the next task?

Reply:
Oh, sorry that's `execution->finishedTaskCount == execution->scheduledTaskCount)`